fastapi/app/api/api_v1/endpoints/images.py

```diff
@@ -0,0 +1,247 @@
+import io
+import json
+import tempfile
+from typing import Optional
+from urllib.parse import unquote
+
+import numpy as np
+import plotly.graph_objects as go
+
+# noinspection PyUnresolvedReferences
+import vtkmodules.vtkInteractionStyle  # noqa
+
+# noinspection PyUnresolvedReferences
+import vtkmodules.vtkRenderingOpenGL2  # noqa
+from PIL import Image
+from starlette.responses import StreamingResponse
+from vtkmodules.vtkCommonCore import vtkFloatArray
+from vtkmodules.vtkCommonCore import vtkIdList
+from vtkmodules.vtkCommonCore import vtkLookupTable
+from vtkmodules.vtkCommonCore import vtkPoints
+from vtkmodules.vtkCommonDataModel import vtkCellArray
+from vtkmodules.vtkCommonDataModel import vtkPolyData
+from vtkmodules.vtkIOImage import vtkPNGWriter
+from vtkmodules.vtkRenderingAnnotation import vtkCubeAxesActor
+from vtkmodules.vtkRenderingAnnotation import vtkScalarBarActor
+from vtkmodules.vtkRenderingCore import vtkActor
+from vtkmodules.vtkRenderingCore import vtkCamera
+from vtkmodules.vtkRenderingCore import vtkPolyDataMapper
+from vtkmodules.vtkRenderingCore import vtkRenderer
+from vtkmodules.vtkRenderingCore import vtkRenderWindow
+from vtkmodules.vtkRenderingCore import vtkTextActor
+from vtkmodules.vtkRenderingCore import vtkWindowToImageFilter
+
+from fastapi import APIRouter
+from fastapi import Header
+
+from .utils import get_girder_client
+from .variables import get_timestep_plot
+
+router = APIRouter()
+
+
+def create_plotly_image(plot_data: dict, format: str, zoom: dict):
+    # The json contains Javascript syntax. Update values for Python
+    for data in plot_data["data"]:
+        data["mode"] = "lines"
+    plot_data["layout"]["autosize"] = True
+    plot_data["layout"]["xaxis"]["automargin"] = True
+    plot_data["layout"]["yaxis"]["automargin"] = True
+    plot_data["layout"]["title"]["x"] = 0.5
+    plot_data["layout"].pop("name", None)
+    plot_data["layout"].pop("frames", None)
+    if zoom:
+        plot_data["layout"]["xaxis"]["range"] = zoom["xAxis"]
+        plot_data["layout"]["yaxis"]["range"] = zoom["yAxis"]
+
+    # Get image as bytes
+    fig = go.Figure(plot_data["data"], plot_data["layout"])
+    return fig.to_image(format, width=500, height=500)
+
+
+def _convert_image(png_image, format: str):
+    img = Image.open(png_image)
+    img = img.convert("RGB")
+    buf = io.BytesIO()
+    img.save(buf, format.upper())
+    img.close()
+    return buf.getvalue()
+
+
+def _mkVtkIdList(it):
+    # Makes a vtkIdList from a Python iterable.
+    vil = vtkIdList()
+    for i in it:
+        vil.InsertNextId(int(i))
+    return vil
+
+
+def create_colormap_image(plot_data: dict, format: str):
+    return
+
+
+def create_mesh_image(plot_data: dict, format: str, zoom: dict):
+    nodes = np.asarray(plot_data["nodes"])
+    connectivity = np.asarray(plot_data["connectivity"])
+    color = np.asarray(plot_data["color"])
```

Comment:
Why aren't these already arrays? Where does the plot_data come from?

Reply:
These are `memoryview` objects, generated [here](https://github.com/bnmajor/adios-dashboard/blob/save-plots/fastapi/app/api/api_v1/endpoints/mesh.py#L6-L27) (the same function that generates the data for serving the image data).

Reply:
Ok, good, so we won't be copying the data.